src/timer.rs

```diff
@@ -0,0 +1,72 @@
+trait Timer {
+    fn start(milliseconds: u32) -> Self;
+    fn has_elapsed(self) -> Option<Self>
+    where
+        Self: Sized;
+}
+
+use libtock_core::syscalls;
+
+// see: https://github.com/tock/tock/blob/master/doc/syscalls/00000_alarm.md
+mod command_nr {
+    pub const _IS_DRIVER_AVAILABLE: usize = 0;
+    pub const GET_CLOCK_FREQUENCY: usize = 1;
+    pub const GET_CLOCK_VALUE: usize = 2;
+    pub const _STOP_ALARM: usize = 3;
+    pub const _SET_ALARM: usize = 4;
+}
+const DRIVER_NUMBER: usize = 0x00000;
+
+lazy_static! {
+    static ref CLOCK_FREQUENCY: usize =
+        syscalls::command(DRIVER_NUMBER, command_nr::GET_CLOCK_FREQUENCY, 0, 0).unwrap_or(0);
+}
+
+struct LibtockAlarmTimer {
+    start_tick: usize,
+    milliseconds: u32,
+}
+
+fn wrapping_sub_u24(lhs: usize, rhs: usize) -> usize {
+    lhs.wrapping_sub(rhs) & 0xffffff
+}
+
+impl Timer for LibtockAlarmTimer {
+    fn start(milliseconds: u32) -> Self {
+        let start_tick =
+            syscalls::command(DRIVER_NUMBER, command_nr::GET_CLOCK_VALUE, 0, 0).unwrap_or(0);
```

Comment:
Here, we can also do the frequency syscall and compute the expiration/end tick. We should add a comment that we assume the clock is 24-bits.

```rust
let delta_tick = frequency * milliseconds / 1000;  // or something against overflow (see below)
assert!(delta_tick < 0x800000);
let end_tick = wrapping_add_u24(start_tick, delta_tick);
```

To compute `delta_tick` without overflow we could do things like:

```rust
// Assume frequency multiple of 100 and milliseconds multiple of 10.
let delta_tick = (frequency / 100) * (milliseconds / 10);
// Assume frequency multiple of 2. We already assert the result to be on 23 bits, so multiplying the result by 500 should fit 32 bits, so the unwrap is already rejecting some of the cases where the delta is too big.
let delta_tick = (frequency / 2).checked_mul(milliseconds).unwrap() / 500;
```

Reply:
We put some thought into how to not overflow in this calculation [here](https://github.com/google/OpenSK/blob/f2496a8e6d71a4e838884996a1c9b62121f87df2/third_party/libtock-drivers/src/timer.rs#L191). Might be cleaner?

Reply:
I thought I wrote about the `u64` computation but probably forgot. Yes that would be an option also (except that `u64` divisions might be very costly on embedded, to be checked).

Note however that the alternative in comments still overflows.

Reply:
Done with attempt to avoid division.

Reply:
Done with attempt to avoid division. PTAL.